### PR TITLE
Add nondet initialisation for statically sized arrays

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/cbmc/array-function-parameters/test.c [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+};
+
+struct Test
+{
+  Node_T *lists[4];
+};
+
+void test(struct Test Test)
+{
+  assert(Test.lists[1]->next);
```
Comment (danpoe) [inline regression/cbmc/array-function-parameters/test.c:21]: We could also check for the other indices 0, 2, and 3, and enable `--bounds-check` and then also access `Test.lists[4]` to check that an out of bounds failure is still reported. In addition we could also set `--max-nondet-tree-depth 2` and then add assertions `assert(Test.lists[<n>]->next->next == NULL)` to check that the array members are initialized with `gen_nondet_init()` just like ordinary variables.

**Review by danpoe — APPROVED** [review]
Looks good to me, though I'd like a slightly more thorough test case

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 6a2767cc54dafe83e452194a1b9118b27ea2841b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92735306
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 51973f36199eae3d8171356522328e20b33aa925).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92736283

**Review by sonodtt — APPROVED** [review]

File: regression/cbmc/array-function-parameters/test.c [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  assert(!Test.lists[1]->next->next);
+  assert(!Test.lists[2]->next->next);
+  assert(!Test.lists[3]->next->next);
+
+  // array elements are not all initialised to the same size
+  assert(Test.lists[0] != Test.lists[1]);
+  assert(Test.lists[1] != Test.lists[2]);
+  assert(Test.lists[2] != Test.lists[3]);
+
+  assert(!Test.lists[4]);
```
Comment (sonodtt) [inline regression/cbmc/array-function-parameters/test.c:36]: Nice to have a comment on this e.g as Dan wrote:
"check that an out of bounds failure is still reported"

File: regression/cbmc/array-function-parameters/test.c (hunk 1 above)
Comment (tautschnig) [reply regression/cbmc/array-function-parameters/test.c:21]: Very much in favour of this.

File: regression/cbmc/array-function-parameters/test.c [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#include <assert.h>
+#include <stdlib.h>
+
+typedef struct Node_T Node_T;
```
Comment (tautschnig) [inline regression/cbmc/array-function-parameters/test.c]: Nit pick: C-land doesn't have camels.

File: scripts/delete_failing_smt2_solver_tests [hunk 4]
```diff
@@ -154,6 +154,7 @@ rm memset1/test.desc
 rm mm_io1/test.desc
 rm no_nondet_static/test.desc
 rm null1/test.desc
+rm array-function-parameters/test.desc
```
Comment (tautschnig) [inline scripts/delete_failing_smt2_solver_tests]: Lexicographic sorting, please.

File: src/ansi-c/c_nondet_symbol_factory.cpp [hunk 5]
```diff
@@ -216,6 +233,25 @@ void symbol_factoryt::gen_nondet_init(
...
 }
 
+void symbol_factoryt::gen_nondet_array_init(
+  code_blockt &assignments,
+  const exprt &expr,
+  std::size_t depth,
+  const recursion_sett &recursion_set)
+{
+  auto const &array_type = to_array_type(expr.type());
+  auto const array_size = numeric_cast_v<mp_integer>(array_type.size());
```
Comment (tautschnig) [inline src/ansi-c/c_nondet_symbol_factory.cpp]: Use `numeric_cast<std::size_t>` and equally use `std::size_t` in the `for`-loop below. If that array is larger than the maximum value representable on the machine we are running on then the loop below wouldn't succeed either...

File: src/ansi-c/c_nondet_symbol_factory.cpp [hunk 6]
```diff
@@ -216,6 +233,25 @@ void symbol_factoryt::gen_nondet_init(
...
+  DATA_INVARIANT(array_size >= 0, "Arrays should have non-negative size");
+  for(auto index = mp_integer(0); index < array_size; ++index)
+  {
+    gen_nondet_init(
+      assignments,
+      index_exprt{expr, from_integer(index, size_type())},
```
Comment (tautschnig) [inline src/ansi-c/c_nondet_symbol_factory.cpp]: Nit pick: use parentheses, not braces. Just because we don't use them otherwise.

**Review by tautschnig — APPROVED** [review]
Just minor issues, but please address them. Hence not merging immediately.

File: regression/cbmc/array-function-parameters/test.c (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/array-function-parameters/test.c]: I'll replace with the cprover naming convention

File: src/ansi-c/c_nondet_symbol_factory.cpp (hunk 6 above)
Comment (hannes-steffenhagen-diffblue) [reply src/ansi-c/c_nondet_symbol_factory.cpp]: How strongly do you feel about this one? I prefer if we started introducing brace initialisation where we can as it doesn't allow narrowing conversions without a cast.

File: regression/cbmc/array-function-parameters/test.c (hunk 1 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/array-function-parameters/test.c:21]: I already did this

**Comment by hannes-steffenhagen-diffblue** [comment]
I've addressed your comments @tautschnig, let's see if CI passes.

File: src/ansi-c/c_nondet_symbol_factory.cpp (hunk 6 above)
Comment (tautschnig) [reply src/ansi-c/c_nondet_symbol_factory.cpp]: Let's solve this via CODING_STANDARD.md, not guerrilla-style.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: fd9187eeee945c9813dcaf4c275591c50863d079).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93395038
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8b305e0de7af6bbaf7d4371a4344f825c3ddb98e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94209577

**Pull request merged**